ostd/src/sync/wait.rs

```diff
@@ -227,10 +215,44 @@ impl Waiter {
         self.waker.do_wait();
     }
 
+    /// Waits until some condition is met or the cancel condition becomes true.
+    ///
+    /// This method will return `Ok(_)` if the condition returns `Some(_)`, and will stop waiting
+    /// if the cancel condition returns `Err(_)`. In this situation, this method will return the `Err(_)`
+    /// generated by the cancel condition.
+    pub fn wait_until_or_cancelled<F, R, FCancel, E>(
+        &self,
+        mut cond: F,
+        cancel_cond: FCancel,
+    ) -> core::result::Result<R, E>
+    where
+        F: FnMut() -> Option<R>,
+        FCancel: Fn() -> core::result::Result<(), E>,
+    {
+        loop {
+            if let Some(res) = cond() {
+                return Ok(res);
+            };
+
+            if let Err(e) = cancel_cond() {
+                // Close the waker and check again to avoid missing a wake event.
+                self.waker.close();
+                return cond().map_or(Err(e), |res| Ok(res));
```

Comment:
[`Option::ok_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or) can be used instead.